ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -252,7 +238,55 @@ private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClie
     private final SchemaRegistryClient schemaRegistryClient;
     private final JoinNode joinNode;
 
-    JoinHelper(final StreamsBuilder builder,
+    protected JoinerFactory(final StreamsBuilder builder,
+                     final KsqlConfig ksqlConfig,
+                     final KafkaTopicClient kafkaTopicClient,
+                     final FunctionRegistry functionRegistry,
+                     final Map<String, Object> props,
+                     final SchemaRegistryClient schemaRegistryClient,
+                     final JoinNode joinNode) {
+
+      this.builder = builder;
+      this.ksqlConfig = ksqlConfig;
+      this.kafkaTopicClient = kafkaTopicClient;
+      this.functionRegistry = functionRegistry;
+      this.props = props;
+      this.schemaRegistryClient = schemaRegistryClient;
+      this.joinNode = joinNode;
+
+    }
+
+
+    public Joiner getJoiner(DataSource.DataSourceType leftType,
+                            DataSource.DataSourceType rightType) {
+
+      if (leftType.isStream() && rightType.isStream()) {
```

Comment:
i'm not that fussed, but could remove this if block if we had a `Map<Pair<DataSource.DataSourceType,DataSource.DataSourceType>, Supplier<Joiner>>` You could build it in the ctor and then this method could become something like:
```
joinerMap.getOrDefault(new Pair<>(leftType, rightType), () -> throw new KsqlException(""));
```

Reply:
Thanks @dguy , I like your suggestion. I have incorporated it! The code looks much nicer now, thanks!